##############################################################################
# File: LTS_DEG_Spreadsheets.R                                               #
# Purpose:                                                                   #
#       Process the DEG tables generated by LTS_DEG_Analysis.R and generate  #
#       corresponding excel spreadsheets.                                    #
#       - For each contrast generate a pairwise spreadsheet                  #
#       - Generate "Method Comparison" spreadsheets comparing QLF and Exact  #
#         test results                                                       #  
#       - Generate "Method Comparison" spreadsheets for selected contrasts   #
#                                                                            #
# Created: September 1, 2020                                                 #
# Author: Adam Faranda                                                       #
#                                                                            #
##############################################################################

# CHANGE THIS DIRECTORY
#setwd('~/Documents/LEC_Time_Series')
setwd("~/Documents/Adam_LEC_Time_Series_DEG_Analysis_3_Oct_2020")
#load("GeneLengthTable.Rdata")
library(dplyr)
library(cluster)
library(reshape2)
library(org.Mm.eg.db)
library(pheatmap)
library(RColorBrewer)
library(EnhancedVolcano)


source('transcriptomic_analysis_scripts/Overlap_Comparison_Functions.R')
source('transcriptomic_analysis_scripts/Excel_Write_Functions.R')

# Set sub-directory paths
wd<-getwd()
rd<-paste(wd, "LTS_DEG_Analysis_Results", sep="/")
sd<-paste(wd, 'transcriptomic_analysis_scripts', sep='/')
# Create directory to store spreadsheets (if it does not yet exist)
if(!dir.exists('LTS_DEG_Spreadsheets'))
{
  dir.create('LTS_DEG_Spreadsheets')
}

# Load In all DEG Tables generated by LTS_DEG_Analysis.R
deg_tables<-list()
for(f in list.files(path=rd, pattern='Test_DEG.csv')){
  fn<-paste(rd,f,sep='/')
  deg_tables[[gsub(".csv", "", f)]]<-read.csv(fn, row.names = 1)
}

######################### Analyze Wildtype Samples ###########################
# DBI Analysis ####

# Select DEG Tables from the DBI analysis
files<-c(
  "DBI_WT24vs0H_Exact_Test_DEG",
  "DBI_WT48vs24H_Exact_Test_DEG",
  "DBI_WT48vs0H_Exact_Test_DEG",
  "DBI_WT24vs0H_QLFTest_DEG",
  "DBI_WT48vs24H_QLFTest_DEG",
  "DBI_WT48vs0H_QLFTest_DEG",
  "DNA_Link_Experiment_1_WT6vs0H_Exact_Test_DEG",
  "DNA_Link_Experiment_1_WT24vs6H_Exact_Test_DEG",
  "DNA_Link_Experiment_1_WT24vs0H_Exact_Test_DEG",
  "DNA_Link_Experiment_1_WT6vs0H_QLFTest_DEG",
  "DNA_Link_Experiment_1_WT24vs6H_QLFTest_DEG",
  "DNA_Link_Experiment_1_WT24vs0H_QLFTest_DEG",
  "DNA_Link_Experiment_2_WT120vs0H_Exact_Test_DEG",
  "DNA_Link_Experiment_2_WT120vs0H_QLFTest_DEG",
  "Global_Wildtype_WT6_QLFTest_DEG",
  "Global_Wildtype_WT24_QLFTest_DEG",
  "Global_Wildtype_WT48_QLFTest_DEG",
  "Global_Wildtype_WT120_QLFTest_DEG"
)

# Nested dictionary with file-specific metadata.
file_meta<-list(
  DBI_WT24vs0H_Exact_Test_DEG=list(
    idc="gene_id", template=paste(
      sd, "DBI_WT24vs0H_Exact_template.xlsx",sep="/"), 
    Group_1="WT_0H_DBI", Group_2="WT_24H_DBI", C1="DBI_WT24vs0H",
    ds="Exact Test: 24 vs 0 Hours PCS in wildtype mice"
  ),
  DBI_WT48vs24H_Exact_Test_DEG=list(
    idc="gene_id", template=paste(
      sd, "DBI_WT48vs24H_Exact_template.xlsx",sep="/"), 
    Group_1="WT_24H_DBI", Group_2="WT_48H_DBI", C1="DBI_WT48vs24H",
    ds="Exact Test: 48 vs 24 Hours PCS in wildtype mice"
  ),
  DBI_WT48vs0H_Exact_Test_DEG=list(
    idc="gene_id", template=paste(
      sd, "DBI_WT48vs0H_Exact_template.xlsx",sep="/"),
    Group_1="WT_0H_DBI", Group_2="WT_48H_DBI", C1="DBI_WT48vs0H",
    ds="Exact Test: 48 vs 0 Hours PCS in wildtype mice"
  ),
  DBI_WT24vs0H_QLFTest_DEG=list(
    idc="gene_id", template=paste(
      sd, "DBI_WT24vs0H_QLF_template.xlsx",sep="/"), 
    Group_1="WT_0H_DBI", Group_2="WT_24H_DBI", C1="DBI_WT24vs0H",
    ds="Quasi Likelihood F-Test: 24 vs 0 Hours PCS in wildtype mice"
  ),
  DBI_WT48vs24H_QLFTest_DEG=list(
    idc="gene_id", template=paste(
      sd, "DBI_WT48vs24H_QLF_template.xlsx",sep="/"), 
    Group_1="WT_24H_DBI", Group_2="WT_48H_DBI", C1="DBI_WT48vs24H",
    ds="Quasi Likelihood F-Test: 48 vs 24 Hours PCS in wildtype mice"
  ),
  DBI_WT48vs0H_QLFTest_DEG=list(
    idc="gene_id", template=paste(
      sd, "DBI_WT48vs0H_QLF_template.xlsx",sep="/"), 
    Group_1="WT_0H_DBI", Group_2="WT_48H_DBI", C1="DBI_WT48vs0H",
    ds="Quasi Likelihood F-Test: 48 vs 0 Hours PCS in wildtype mice"
  ),
  DNA_Link_Experiment_1_WT6vs0H_Exact_Test_DEG=list(
    idc="gene_id", template=paste(
      sd, "DNA1_WT6vs0H_Exact_template.xlsx",sep="/"), 
    Group_1="WT_0H_DNA1", Group_2="WT_6H_DNA1", C1="DNA1_WT24vs0H",
    ds="Exact Test: 6 vs 0 Hours PCS in wildtype mice"
  ),
  DNA_Link_Experiment_1_WT24vs6H_Exact_Test_DEG=list(
    idc="gene_id", template=paste(
      sd, "DNA1_WT24vs6H_Exact_template.xlsx",sep="/"), 
    Group_1="WT_6H_DNA1", Group_2="WT_24H_DNA1", C1="DNA1_WT24vs6H",
    ds="Exact Test: 24 vs 6 Hours PCS in wildtype mice"
  ),
  DNA_Link_Experiment_1_WT24vs0H_Exact_Test_DEG=list(
    idc="gene_id", template=paste(
      sd, "DNA1_WT24vs0H_Exact_template.xlsx",sep="/"), 
    Group_1="WT_0H_DNA1", Group_2="WT_24H_DNA1", C1="DNA1_WT24vs0H",
    ds="Exact Test: 24 vs 0 Hours PCS in wildtype mice"
  ),
  DNA_Link_Experiment_1_WT6vs0H_QLFTest_DEG=list(
    idc="gene_id", template=paste(
      sd, "DNA1_WT6vs0H_QLF_template.xlsx",sep="/"), 
    Group_1="WT_0H_DNA1", Group_2="WT_6H_DNA1", C1="DNA1_WT24vs0H",
    ds="Quasi Likelihood F-Test: 6 vs 0 Hours PCS in wildtype mice"
  ),
  DNA_Link_Experiment_1_WT24vs6H_QLFTest_DEG=list(
    idc="gene_id", template=paste(
      sd, "DNA1_WT24vs6H_QLF_template.xlsx",sep="/"), 
    Group_1="WT_6H_DNA1", Group_2="WT_24H_DNA1", C1="DNA1_WT24vs6H",
    ds="Quasi Likelihood F-Test: 24 vs 6 Hours PCS in wildtype mice"
  ),
  DNA_Link_Experiment_1_WT24vs0H_QLFTest_DEG=list(
    idc="gene_id", template=paste(
      sd, "DNA1_WT24vs0H_QLF_template.xlsx",sep="/"), 
    Group_1="WT_0H_DNA1", Group_2="WT_24H_DNA1", C1="DNA1_WT24vs0H",
    ds="Quasi Likelihood F-Test: 24 vs 0 Hours PCS in wildtype mice"
  ),
  DNA_Link_Experiment_2_WT120vs0H_Exact_Test_DEG=list(
    idc="gene_id", template=paste(
      sd, "DNA2_WT120vs0H_Exact_template.xlsx",sep="/"),
    Group_1="WT_0H_DNA2", Group_2="WT_120H_DNA2", C1="DNA2_WT120vs0H",
    ds="Exact Test: 120 vs 0 Hours PCS in wildtype mice"
  ),
  DNA_Link_Experiment_2_WT120vs0H_QLFTest_DEG=list(
    idc="gene_id", template=paste(
      sd, "DNA2_WT120vs0H_QLF_template.xlsx",sep="/"),
    Group_1="WT_0H_DNA2", Group_2="WT_120H_DNA2", C1="DNA2_WT120vs0H",
    ds="Quasi Likelihood F-Test: 120 vs 0 Hours PCS in wildtype mice"
  ),
  Global_Wildtype_WT6_QLFTest_DEG=list(
    idc="gene_id", template=paste(sd, "Global_WT6_template.xlsx",sep="/"),
    Group_1="0H", Group_2="6H", C1="DNA2_WT120vs0H",
    ds="Quasi Likelihood F-Test: 6 vs 0 Hours PCS in wildtype mice"
  ),
  Global_Wildtype_WT24_QLFTest_DEG=list(
    idc="gene_id", template=paste(sd, "Global_WT24_template.xlsx",sep="/"), 
    Group_1="0H", Group_2="24H", C1="ALL_6Hvs0H",
    ds="Quasi Likelihood F-Test: 24 vs 0 Hours PCS in wildtype mice"
  ),
  Global_Wildtype_WT48_QLFTest_DEG=list(
    idc="gene_id", template=paste(sd, "Global_WT48_template.xlsx",sep="/"), 
    Group_1="0H", Group_2="48H", C1="ALL_WT120vs0H",
    ds="Quasi Likelihood F-Test: 48 vs 0 Hours PCS in wildtype mice"
  ),
  Global_Wildtype_WT120_QLFTest_DEG=list(
    idc="gene_id", template=paste(sd, "Global_WT120_template.xlsx",sep="/"), 
    Group_1="0H", Group_2="6H", C1="ALL_WT120vs0H",
    ds="Quasi Likelihood F-Test: 120 vs 0 Hours PCS in wildtype mice"
  )
)

# Iterate over DEG Tables and populate DEG Spreadsheets. 
for(f in files){
  print(
    paste(
      f, "Rows:",nrow(deg_tables[[f]]),
      "Cols", ncol(deg_tables[[f]])
    )
  )
  dg1<-deg_tables[[f]]
  if(grepl("Global", f)){
    names(dg1)<-gsub("^X","",names(dg1))
  }
  
  a1<-grep(
    paste("^",file_meta[[f]][['Group_1']],"$",sep=""),
    names(dg1)
  )
  a2<-grep(
    paste("^",file_meta[[f]][['Group_2']],"$",sep=""),
    names(dg1)
  )
  Avg1<-paste(file_meta[[f]][['Group_1']], "FPKM",sep="_")
  Avg2<-paste(file_meta[[f]][['Group_2']], "FPKM",sep="_")
  names(dg1)[a1]<-Avg1
  names(dg1)[a2]<-Avg2
  dg1<-dg1[,c(
    "gene_id", "SYMBOL", "coding_length",
    "DESCRIPTION", "logFC", "PValue", "FDR", 
    Avg1, Avg2
  )]
  print(names(dg1))
  print(
    degSummary(
      dg1,
      Avg1 = file_meta[[f]][['Group_1']],
      Avg2 = file_meta[[f]][['Group_2']]
    )
  )
  
  dg1$Group_1 <- file_meta[[f]][['Group_1']]
  dg1$Group_2 <- file_meta[[f]][['Group_2']]
  
  createDEGSpreadSheet(
    C1 = file_meta[[f]][['C1']],     # Name of the contrast
    dg1 = dg1,                       # Data Set for the contrast
    dg1.bioFun = bioSigRNASeq,       # Biological significance filter for dg1
    dg1.fdr = "FDR",                 # Statistic used to filter genes for dg1
    dg1.lfc = "logFC",               # Column in dg1 with log Fold Changes
    dg1.Avg1 = Avg1,               # Column in dg1 with average value for Group_1
    dg1.Avg2 = Avg2,               # Column in dg1 with average value for Group_2
    dg1.me = 2,                      # Min. expression for dg1.bioFun
    dg1.x = 23,                      # row number, corner of dg1 Summary table
    dg1.y = 2,                       # col number, corner of dg1 Summary table
    dg1.ds = file_meta[[f]][['ds']], # short description for contrast C1 (dg1)
    template = file_meta[[f]][['template']],
    descPageName="Data Description", # Name of sheet to write summary tables
    wb = NULL,                       # Optionally pass a workbook object instead.
    pref = "PCO" ,                      # Prefix for output file.
    fname=paste(
      wd,'/LTS_DEG_Spreadsheets/',
      f,'.xlsx',sep=''  
    ),                               # Manually specify an output file name
    use_lfc = FALSE,                 # Whether to use logFC or Fold_Change,
    cols=c(
      "gene_id",
      "SYMBOL",
      "coding_length",
      "DESCRIPTION", 
      "logFC", 
      "PValue", 
      "FDR", 
      Avg1, 
      Avg2
    )
  )
}


##### EOF #####
