##############################################################################
# File: Runx1_Targets_Time_Series.R
# Purpose: Analyze the changes that occur in Runx1 targets in the 
#          wildtype time series. Tables of potential Runx1 targets were
#          were retrieved from the TRRUST database on October 28, 2020.
#          This script generates time series plots for the Runx1 
#          targets identified by TRRUST that were significantly DE at any
#          interval in the wildtype time series. Files generated by this
#          script (plots and tables) have the prefix "R1T" 
#         
# Created: October 28, 2020
# Author: Adam Faranda
#
##############################################################################



# CHANGE THIS DIRECTORY
# setwd('~/Documents/LEC_Time_Series')
setwd("~/Documents/Adam_LEC_Time_Series_DEG_Analysis_3_Oct_2020")
#load("GeneLengthTable.Rdata")
library(dplyr)
library(cluster)
library(reshape2)
library(pheatmap)
library(RColorBrewer)
library(EnhancedVolcano)
library(pheatmap)
library(tidyr)

wd<-getwd()

source('transcriptomic_analysis_scripts/BuildDataMatrix.R')
source('transcriptomic_analysis_scripts/PreprocessingFunctions.R')
source('transcriptomic_analysis_scripts/PrincipalComponents.R')
source('transcriptomic_analysis_scripts/ClusteringFunctions.R')
source('transcriptomic_analysis_scripts/Overlap_Comparison_Functions.R')

############################ Load in Data Files ##############################

### FPKM Matrix
fpkm<-read.csv(
  "LTS_DEG_Analysis_results/Global_Wildtype_QLFTest_FPKM_Matrix.csv", 
  row.names=1
)

### Differential Expression stats
deg<-read.csv(
  "LTS_DEG_Analysis_results/Global_Wildtype_QLFTest_DEG.csv",
  row.names=1
)
names(deg)<-gsub("^X", "", names(deg))


### Sample Labels & Factors
dl<-paste(
  wd,
  "LEC_Time_Series_HTSeq_Counts", sep=""
)
ft<-hc_getFileTable(
  dirList=dl, filename = "HTSeq_GeneCounts_All.csv"
) %>% filter(genotype=="WT")

ft$sample<-paste(
  ft$genotype, 
  paste(ft$hours.pcs,"H",sep=''), 
  ft$batch,c(7:15, 22:36), sep="_"
)
row.names(ft)<-ft$sample

ft$interval<-droplevels(
  factor(
    paste(ft$hours.pcs, 'H', sep=''),
    levels = c('0H', '6H', '24H','48H', '120H')
  )
)
ft$batch<-droplevels(
  factor(
    ft$batch, 
    levels=c("DBI", "DNA1", "DNA2")
  )
)

### Runx1 Targets
runx1.hs<-read.table(
  paste(
    wd,
    "/LEC_Time_Series_HTSeq_Counts/RUNX1_targets.human.tsv",
    sep=""
  )
)

## Quick and dirty Human Mouse conversion
runx1.hs$V2<-paste(
  substr(runx1.hs$V2,1,1),
  tolower(substr(runx1.hs$V2,2,nchar(runx1.hs$V2))),
  sep=""
)

runx1.mm<-read.table(
  paste(
    wd,
    "/LEC_Time_Series_HTSeq_Counts/Runx1_targets.mouse.tsv",
    sep=""
  )
)
### Build a vector of all Runx1 targets
tgt<-union(runx1.hs$V2, runx1.mm$V2)

### Build Vector of ensembl gene_id's corresponding to Runx1 targets
### with significant differential expression in the Global model
sig.tgt<-deg %>% filter(FDR < 0.05 & SYMBOL %in% tgt) %>% pull(gene_id)
names(sig.tgt) <- deg[sig.tgt,"SYMBOL"]

### Iterate over significant targets, generate plots and populate a 
### table of measurements
plt.all<-data.frame()
for (g in names(sig.tgt)){
  if(g %in% deg$SYMBOL){
    plt<-inner_join(
      ft, 
      reshape2::melt(fpkm %>% filter(gene_id == sig.tgt[g])) %>%
        filter(variable != "coding_length") %>%
        dplyr::select(sample=variable, FPKM=value),
      by="sample"
    )
    plt$gene_symbol<-g
    plt$ensembl_id<-sig.tgt[g]
    ggplot(plt, aes(x=hours.pcs, y=FPKM, group=interval)) + 
      geom_boxplot() + ggtitle(g)
    ggsave(
      filename=paste(
        "LTS_DEG_Analysis_results/R1T_",g,
        "_fpkm_box.png", sep=""
      ), width=6, height=2
    )
    
    ggplot(plt, aes(x=hours.pcs, y=FPKM, group=interval)) + 
      geom_point(aes(color=batch)) + ggtitle(g)
    ggsave(
      filename=paste(
        "LTS_DEG_Analysis_results/R1T_",g,
        "_fpkm_pnt.png", sep=""
      ), width=6, height=2
    )
    plt.all<-bind_rows(plt.all, plt)
  }
}

head(plt.all)
write.csv(
  plt.all %>% 
    group_by(gene_symbol, interval) %>%
    summarize(Avg_FPKM = mean(FPKM)) %>%
    mutate(logFC_0H=log2(Avg_FPKM/Avg_FPKM[1])) %>%
    filter(interval != "0H") %>%
    mutate(
      is.max=abs(logFC_0H) == max(abs(logFC_0H)),
      is.min=abs(logFC_0H) == min(abs(logFC_0H))
    ) %>% group_by(interval) %>%
    summarise(MaxResponse=sum(is.max), MinResoponse=sum(is.min)),
  "LTS_DEG_Analysis_results/R1T_Max_Min_Response_Table.csv"
)


ggplot(plt.all, aes(x=interval, y=FPKM, group=interval)) + 
  geom_point(aes(color=batch)) + 
  ggtitle("Runx1 targets in LEC Time Series") +
  facet_wrap( ~gene_symbol, scales="free")+
  theme(axis.text.x = element_text(angle=45))
ggsave(
  filename=paste(
    "LTS_DEG_Analysis_results/R1T_Runx1_target_LTS_PanelPlot.png",sep=""
  ), width=9, height=12
)



annot<-plt.all %>% 
  select( sample, interval) %>% distinct()
row.names(annot)<-annot$sample

x<-plt.all %>%
  select(sample, gene_symbol, FPKM) %>%
  mutate(
    FPKM = log(FPKM + min(plt.all %>% filter(FPKM > 0) %>% pull(FPKM)),2)
  ) %>%
  group_by(gene_symbol) %>%
  mutate(FPKM = (FPKM - mean(FPKM))) %>%
  mutate(FPKM = (FPKM / sd(FPKM))) %>%
  pivot_wider(id_cols=gene_symbol, names_from=sample, values_from=FPKM) %>%
  as.data.frame()

row.names(x)<-x$gene_symbol
x<-as.data.frame(x[,setdiff(1:ncol(x), grep("gene_symbol", names(x)))])

## Clustering on log transformed, scaled FPKM. The minimum Non
## Negative value is added to all measurements before log transformation. 

pheatmap(
    x, scale ="none", annotation_col = annot[2]
)


write.table(
  data.frame(gene_id=row.names(x),x),
  file="LTS_DEG_Analysis_results/R1T_Scaled_Clustering_Matrix.txt",
  sep="\t", row.names = F, quote = F
)

ggsave(
  "LTS_DEG_Analysis_results/R1T_Runx1_target_LTS_Heatmap.png",
  pheatmap(
    log(
      x + min(plt.all %>% filter(FPKM > 0) %>% pull(FPKM))
    ), scale ="row", annotation_col = annot[2]
  )
)
